modules/packages/Sort.chpl

```diff
@@ -163,46 +323,49 @@ proc InsertionSort(Data: [?Dom] ?eltType, doublecheck=false, param reverse=false
    :type doublecheck: `bool`
    :arg reverse: Sort in reverse numerical order
    :type reverse: `bool`
+   :arg comparator: Record that redefines the comparison mechanism with one of
+      the methods: ``comparator.key(a)`` or ``comparator.compare(a,b)``
 
  */
-proc MergeSort(Data: [?Dom], minlen=16, doublecheck=false, param reverse=false) where Dom.rank == 1 {
-  _MergeSort(Data, minlen, reverse=reverse);
-  if (doublecheck) then VerifySort(Data, "MergeSort", reverse);
+proc MergeSort(Data: [?Dom] ?eltType, minlen=16, doublecheck=false, param reverse=false, comparator:?rec=emptyRecord) where Dom.rank == 1 && isRecord(rec) {
+  chpl_check_comparator(comparator, Data(Dom.dim(1).low));
+  _MergeSort(Data, minlen, reverse=reverse, comparator=comparator);
+  if (doublecheck) then VerifySort(Data, "MergeSort", reverse, comparator);
 }
 
 
-private proc _MergeSort(Data: [?Dom], minlen=16, param reverse=false) where Dom.rank == 1 {
+private proc _MergeSort(Data: [?Dom], minlen=16, param reverse=false, comparator:?rec=emptyRecord) where Dom.rank == 1 && isRecord(rec) {
   const lo = Dom.dim(1).low;
   const hi = Dom.dim(1).high;
   if hi-lo < minlen {
-    InsertionSort(Data, reverse=reverse);
+    InsertionSort(Data, reverse=reverse, comparator=comparator);
     return;
   }
   const mid = (hi-lo)/2+lo;
   var A1 = Data[lo..mid];
   var A2 = Data[mid+1..hi];
   cobegin {
-    { _MergeSort(A1, minlen, reverse=reverse); }
-    { _MergeSort(A2, minlen, reverse=reverse); }
+    { _MergeSort(A1, minlen, reverse=reverse, comparator=comparator); }
+    { _MergeSort(A2, minlen, reverse=reverse, comparator=comparator); }
   }
 
-  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse)) do a = _a;
+  for (a, _a) in zip(Data[lo..hi], _MergeIterator(A1, A2, reverse=reverse, comparator=comparator)) do a = _a;
```

Comment:
I have trouble believing that this iterator-based implementation of MergeSort on arrays is anything other than a test of a language feature.
I'd expect MergeSort to outperform QuickSort if implemented well. I think this iterator is just adding overhead. It would be nice to add a comment indicating it's adding overhead (iterators with more than 1 yield can't be inlined) at not much benefit since the iterator is only called once.